include/madwf_ml.h

```diff
@@ -0,0 +1,372 @@
+#pragma once
+
+#include <vector>
+#include <random>
+#include <unordered_map>
+
+#include <color_spinor_field.h>
+#include <blas_quda.h>
+#include <madwf_transfer.h>
+#include <polynomial.h>
+
+#include <random_quda.h>
+#include <tune_quda.h>
+
+#include <device_vector.h>
+
+namespace quda
+{
+
+  struct MadwfAcc {
+
+    using transfer_float = float;
+
+    static constexpr madwf_ml::transfer_5D_type transfer_type = madwf_ml::transfer_5D_type::Spin;
+
+    // The parameters to be trained.
+    using device_container = device_vector<transfer_float>;
+    device_container device_param;
+
+    // The diagonal component to suppress the zero modes.
+    double mu;
+    int Ls_base;
+
+    int null_miniter;
+    double null_tol;
+    int train_maxiter;
+
+    // persistent buffers for reuse.
+    ColorSpinorField *forward_tmp = nullptr;
+    ColorSpinorField *backward_tmp = nullptr;
+
+    bool param_load;
+    bool param_save;
+
+    char param_infile[256];
+    char param_outfile[256];
+
+    // Has device_param been trained?
+    bool trained = false;
+
+    QudaPrecision prec_precondition;
+
+    static std::unordered_map<std::string, std::vector<transfer_float>> host_training_param_cache; // empty map
+
+    MadwfAcc(const SolverParam &solve_param) :
+      mu(solve_param.madwf_diagonal_suppressor),
+      Ls_base(solve_param.madwf_ls),
+      null_miniter(solve_param.madwf_null_miniter),
+      null_tol(solve_param.madwf_null_tol),
+      train_maxiter(solve_param.madwf_train_maxiter),
+      param_load(solve_param.madwf_param_load == QUDA_BOOLEAN_TRUE),
+      param_save(solve_param.madwf_param_save == QUDA_BOOLEAN_TRUE),
+      prec_precondition(solve_param.precision_precondition)
+    {
+      strcpy(param_infile, solve_param.madwf_param_infile);
+      strcpy(param_outfile, solve_param.madwf_param_outfile);
+
+      printfQuda("Launching MADWF accelerator ... \n");
```

Comment:
The `printfQuda` calls in this file should be guarded behind `getVerbosity()`, so that is the user requests silent, nothing is printed.

Reply:
https://github.com/lattice/quda/pull/1061/commits/96926239a741adb4fa5c8bab17dc6714e094af3b